WordPress/Classes/Services/MediaImportService.swift

```diff
@@ -20,55 +20,254 @@ open class MediaImportService: LocalCoreDataService {
+    ///   - completion: a block that will be invoked when the media is created, on success it will return a valid Media
+    ///         object, on failure it will return a nil Media and an error object with the details.
+    @objc
+    @discardableResult
+    func createMedia(
+        with exportable: ExportableAsset,
+        blog: Blog,
+        post: AbstractPost?,
+        receiveUpdate: ((Media) -> Void)?,
+        thumbnailCallback: ((Media, URL) -> Void)?,
+        completion: @escaping (Media?, Error?) -> Void
+    ) -> Progress {
+        let createProgress = Progress.discreteProgress(totalUnitCount: 1)
+        coreDataStack.performAndSave({ context in
+            let media = try self.createMedia(with: exportable, blogObjectID: blog.objectID, postObjectID: post?.objectID, in: context)
+            try context.obtainPermanentIDs(for: [media])
+            return media.objectID
+        }, completion: { (result: Result<NSManagedObjectID, Error>) in
+            let transformed = result.flatMap { mediaObjectID in
+                Result {
+                    (
+                        try self.coreDataStack.mainContext.existingObject(with: mediaObjectID) as! Media,
+                        try self.coreDataStack.mainContext.existingObject(with: blog.objectID) as! Blog
+                    )
+                }
+            }
+            switch transformed {
+            case let .success((media, blog)):
+                let progress = self.import(exportable, to: media, blog: blog, thumbnailCallback: thumbnailCallback) {
+                    switch $0 {
+                    case let .success(media):
+                        completion(media, nil)
+                    case let .failure(error):
+                        completion(nil, error)
+                    }
+                }
+                createProgress.addChild(progress, withPendingUnitCount: 1)
+                receiveUpdate?(media)
+            case let .failure(error):
+                completion(nil, error)
+            }
+        }, on: .main)
+        return createProgress
+    }
+
+    private func createMedia(with exportable: ExportableAsset, blogObjectID: NSManagedObjectID, postObjectID: NSManagedObjectID?, in context: NSManagedObjectContext) throws -> Media {
+        let blogInContext = try context.existingObject(with: blogObjectID) as! Blog
+        let postInContext = try postObjectID.flatMap(context.existingObject(with:)) as? AbstractPost
+
+        let media = postInContext.flatMap(Media.makeMedia(post:)) ?? Media.makeMedia(blog: blogInContext)
+        media.mediaType = exportable.assetMediaType
+        media.remoteStatus = .processing
+        return media
+    }
+
+    private func `import`(_ exportable: ExportableAsset, to media: Media, blog: Blog,
+                          thumbnailCallback: ((Media, URL) -> Void)?,
+                          completion: @escaping (Result<Media, Error>) -> Void) -> Progress {
```

Comment:
Nitpick: Just for consistency

```suggestion
    private func `import`(
        _ exportable: ExportableAsset, 
        to media: Media,
        blog: Blog,
        thumbnailCallback: ((Media, URL) -> Void)?,
        completion: @escaping (Result<Media, Error>) -> Void
    ) -> Progress {
```